cpp/tests/utilities/column_utilities.cu

```diff
@@ -46,78 +47,310 @@
+  //
+  // result = [6, 1, 11, 1, 1]
+  //
+  auto validity_iter = cudf::detail::make_counting_transform_iterator(
+    0,
+    [row_indices = row_indices.begin<size_type>(),
+     validity    = c.null_mask(),
+     offset      = c.offset()] __device__(int index) {
+      auto const true_index = row_indices[index] + offset;
+      return !validity || cudf::bit_is_set(validity, true_index) ? 1 : 0;
+    });
+  auto output_row_iter = cudf::detail::make_counting_transform_iterator(
+    0,
+    [row_indices  = row_indices.begin<size_type>(),
+     offsets      = c.offsets().begin<offset_type>(),
+     offset       = c.offset(),
+     first_offset = cudf::detail::get_value<offset_type>(
+       c.offsets(), c.offset(), rmm::cuda_stream_default)] __device__(int index) {
+      auto const true_index = row_indices[index] + offset;
+      return offsets[true_index] - first_offset;
+    });
+  thrust::scatter_if(rmm::exec_policy(),
+                     output_row_iter,
+                     output_row_iter + row_indices.size(),
+                     output_row_start->view().begin<size_type>(),
+                     validity_iter,
+                     result->mutable_view().begin<size_type>());
+
+  // generate keys for each output row
+  //
+  // result = [1, 1, 2, 2, 2]
+  //
+  auto keys =
+    cudf::make_fixed_width_column(data_type{type_id::INT32}, output_size, mask_state::UNALLOCATED);
+  thrust::generate(rmm::exec_policy(),
+                   keys->mutable_view().begin<size_type>(),
+                   keys->mutable_view().end<size_type>(),
+                   [] __device__() { return 0; });
+  thrust::scatter_if(rmm::exec_policy(),
+                     validity_iter,
+                     validity_iter + row_indices.size(),
+                     output_row_start->view().begin<size_type>(),
+                     validity_iter,
+                     keys->mutable_view().begin<size_type>());
+  thrust::inclusive_scan(rmm::exec_policy(),
+                         keys->view().begin<size_type>(),
+                         keys->view().end<size_type>(),
+                         keys->mutable_view().begin<size_type>());
+
+  // scan by key to generate final child row indices.
+  // input
+  //    result = [6, 1, 11, 1, 1]
+  //    keys   = [1, 1, 2,  2, 2]
+  //
+  // output
+  //    result = [6, 7, 11, 12, 13]
+  //
+  thrust::inclusive_scan_by_key(rmm::exec_policy(),
+                                keys->view().begin<size_type>(),
+                                keys->view().end<size_type>(),
+                                result->view().begin<size_type>(),
+                                result->mutable_view().begin<size_type>());
+  return result;
+}
+
+#define PROP_EXPECT_EQ(a, b, _msg) \
+  do {                             \
+    if (throw_on_fail) {           \
+      CUDF_EXPECTS(a == b, _msg);  \
+    } else {                       \
+      EXPECT_EQ(a, b);             \
+    }                              \
+  } while (0)
```

Comment:
I'm unfamiliar with this idiom. Why a ` do {} while (0);` loop instead of just the `if` block?

Reply:
Here it is https://stackoverflow.com/a/2314201 :smile:

Reply:
It also adds an extra scope that gives you a little bit more insulation for local variables, etc.